cudax/include/cuda/experimental/__memory_resource/any_resource.cuh

```diff
@@ -42,17 +42,150 @@
 #include <cuda/std/__concepts/__concept_macros.h>
 #include <cuda/std/__concepts/_One_of.h>
 #include <cuda/std/__concepts/all_of.h>
+#include <cuda/std/__type_traits/is_copy_constructible.h>
 #include <cuda/std/__type_traits/is_nothrow_constructible.h>
 #include <cuda/std/__type_traits/is_same.h>
 #include <cuda/std/__type_traits/remove_cvref.h>
 #include <cuda/std/__utility/exchange.h>
 #include <cuda/std/__utility/forward.h>
+#include <cuda/std/__utility/in_place.h>
+#include <cuda/std/atomic>
 
 namespace cuda::experimental::mr
 {
 template <class _Ty, class _Uy = _CUDA_VSTD::remove_cvref_t<_Ty>>
 _LIBCUDACXX_INLINE_VAR constexpr bool __is_basic_any_resource = false;
 
+template <class _Resource,
+          _CUDA_VMR::_AllocType =
+            _CUDA_VMR::async_resource<_Resource> ? _CUDA_VMR::_AllocType::_Async : _CUDA_VMR::_AllocType::_Default>
+struct _Shared_resource
+{
+  static_assert(_CUDA_VMR::resource<_Resource>, "");
+
+  template <class... _Args>
+  explicit _Shared_resource(_Args&&... __args)
+      : __control_block(new _Control_block{_Resource{_CUDA_VSTD::forward<_Args>(__args)...}, 1})
+  {}
+
+  _Shared_resource(const _Shared_resource& __other) noexcept
+      : __control_block(__other.__control_block)
+  {
+    __control_block->__ref_count.fetch_add(1, _CUDA_VSTD::memory_order_relaxed);
+  }
+
+  _Shared_resource(_Shared_resource&& __other) noexcept
+      : __control_block(_CUDA_VSTD::exchange(__other.__control_block, nullptr))
+  {}
+
+  ~_Shared_resource()
+  {
+    if (__control_block && __control_block->__ref_count.fetch_sub(1, _CUDA_VSTD::memory_order_acq_rel) == 1)
+    {
+      delete __control_block;
+    }
+  }
+
+  _Shared_resource& operator=(const _Shared_resource& __other) noexcept
+  {
+    if (this != &__other)
+    {
+      _Shared_resource(__other).swap(*this);
+    }
+
+    return *this;
+  }
+
+  _Shared_resource& operator=(_Shared_resource&& __other) noexcept
+  {
+    if (this != &__other)
+    {
+      _Shared_resource(_CUDA_VSTD::move(__other)).swap(*this);
+    }
+
+    return *this;
+  }
+
+  void swap(_Shared_resource& __other) noexcept
+  {
+    _CUDA_VSTD::swap(__control_block, __other.__control_block);
+  }
```

Comment:
Nitpick: Can we move that down after the allocate / deallocate function so that it is logically grouped?